llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -352,126 +355,71 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
     EntryPointsGroups[GLOBAL_SCOPE_NAME] = {};
 }
 
-enum HasAssertStatus { No_Assert, Assert, Assert_Indirect };
-
-// Go through function call graph searching for assert call.
-HasAssertStatus hasAssertInFunctionCallGraph(const Function *Func) {
-  // Map holds the info about assertions in already examined functions:
-  // true  - if there is an assertion in underlying functions,
-  // false - if there are definetely no assertions in underlying functions.
-  static std::map<const Function *, bool> hasAssertionInCallGraphMap;
-  std::vector<const Function *> FuncCallStack;
-
-  static std::vector<const Function *> isIndirectlyCalledInGraph;
-
-  std::vector<const Function *> Workstack;
-  Workstack.push_back(Func);
-
-  while (!Workstack.empty()) {
-    const Function *F = Workstack.back();
-    Workstack.pop_back();
-    if (F != Func)
-      FuncCallStack.push_back(F);
-
-    bool HasIndirectlyCalledAttr = false;
-    if (std::find(isIndirectlyCalledInGraph.begin(),
-                  isIndirectlyCalledInGraph.end(),
-                  F) != isIndirectlyCalledInGraph.end())
-      HasIndirectlyCalledAttr = true;
-    else if (F->hasFnAttribute("referenced-indirectly")) {
-      HasIndirectlyCalledAttr = true;
-      isIndirectlyCalledInGraph.push_back(F);
-    }
-
-    bool IsLeaf = true;
-    for (const auto &I : instructions(F)) {
-      if (!isa<CallBase>(&I))
-        continue;
+// This function traverses over reversed call graph by BFS algorithm.
+// It means that an edge links some function @func with functions
+// which contain call of function @func. It starts from
+// @StartingFunction and lifts up until it reach all reachable functions
+// or it reaches some function containing "referenced-indirectly" attribute.
+// If it reaches "referenced-indirectly" attribute than it returns an empty
+// Optional.
+// Otherwise, it returns an Optional containing a list of reached
+// SPIR kernel function's names.
+Optional<std::vector<StringRef>>
+TraverseCGToFindSPIRKernels(const Function *StartingFunction) {
+  std::queue<const Function *> FunctionsToVisit;
+  std::unordered_set<const Function *> VisitedFunctions;
+  FunctionsToVisit.push(StartingFunction);
+  std::vector<StringRef> KernelNames;
+
+  while (!FunctionsToVisit.empty()) {
+    const Function *F = FunctionsToVisit.front();
+    FunctionsToVisit.pop();
+
+    auto InsertionResult = VisitedFunctions.insert(F);
+    // It is possible that we insert some particular function several
+    // times in functionsToVisit queue.
+    if (!InsertionResult.second)
+      continue;
 
-      const Function *CF = cast<CallBase>(&I)->getCalledFunction();
-      if (!CF)
+    for (const auto *U : F->users()) {
+      const Instruction *I = cast<const Instruction>(U);
+      const Function *ParentF = I->getFunction();
```

Comment:
I'm concerned that there could be other uses of function, which are not instructions, but [`ConstantExpr`](https://llvm.org/doxygen/classllvm_1_1ConstantExpr.html). To generate such LLVM IR we can try to cast one function pointer type into another before doing a call. There is an example of such LLVM IR in [the translator repo](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/test/constexpr_vector.ll)

Another concern is that theoretically, an Instruction can be located in module scope,  i.e. outside of a function. Therefore, `I->getFunction()` will return `nullptr`. There is an example of such LLVM IR in [the translator repo](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/test/transcoding/SPV_INTEL_function_pointers/const-function-pointer.ll) as well. Note: I don't have good idea about how to generate such LLVM IR out of a SYCL app if you want to replicate it yourself.

Reply:
Added `dyn_cast` to `CallInst`.

Reply:
Please note that even though you have fixed a nullptr dereference, functionally we still have an issue, i.e. we can say that there are no assert uses whilst there are through more complex usages of function pointers.

That is probably a minor problem at the moment, because function pointers are not widely used, but we should leave a `FIXME` here, at least, mentioning the use cases we don't analyze properly.

Reply:
Currently, there is a kernel `empty_kernel` in assert-indirect-with-split-2 which checks that traversing doesn't go in such cases.